rust/foxglove/src/websocket.rs

```diff
@@ -780,14 +802,87 @@ impl ConnectedClient {
+        if is_subscribed {
+            tracing::info!(
+                "Client {} is already subscribed to connection graph updates",
+                self.addr
+            );
+            return;
+        }
+
+        {
+            let mut subscriber_count = server.connection_graph_subscriber_count.lock();
+            let is_first_subscriber = *subscriber_count == 0;
+            *subscriber_count += 1;
+
+            // We hold the lock over the call to the listener so that subscribe and unsubscribe
+            // calls are correctly ordered relative to each other.
+            if is_first_subscriber {
+                if let Some(listener) = server.listener.as_ref() {
+                    listener.on_connection_graph_subscribe();
+                }
+            }
+        }
+
+        // We hold the connection_graph lock over updating self.subscribed_to_connection_graph
+        // and sending the initial update message, so it's synchronized with unsubscribe and
+        // with server.connection_graph_update.
+        let mut connection_graph = server.connection_graph.lock();
+        // Take the graph and replace it with an empty default
+        let current_graph = std::mem::take(&mut *connection_graph);
+        // Update the empty default with the current graph, setting it back to where it was,
+        // and generating the full diff in the process.
+        let json_diff = connection_graph.update(current_graph);
+        // Send the full diff to the client as the starting state
+        self.send_control_msg(Message::text(json_diff));
+        self.subscribed_to_connection_graph.store(true, Relaxed);
+    }
+
+    fn on_connection_graph_unsubscribe(&self, server: Arc<Server>) {
+        if !server.capabilities.contains(&Capability::ConnectionGraph) {
+            self.send_error("Server does not support connection graph capability".to_string());
+            return;
+        }
+
+        let is_subscribed = self.subscribed_to_connection_graph.load(Relaxed);
+        if !is_subscribed {
+            self.send_error("Client is not subscribed to connection graph updates".to_string());
+            return;
+        }
+
+        {
+            let mut subscriber_count = server.connection_graph_subscriber_count.lock();
+            *subscriber_count -= 1;
+
+            if *subscriber_count == 0 {
+                if let Some(listener) = server.listener.as_ref() {
+                    listener.on_connection_graph_unsubscribe();
+                }
+            }
+        }
+
+        // Acquire the lock to sychronize with subscribe and with server.connection_graph_update.
+        let _guard = server.connection_graph.lock();
+        self.subscribed_to_connection_graph.store(false, Relaxed);
+    }
+
     /// Send an ad hoc error status message to the client, with the given message.
     fn send_error(&self, message: String) {
+        tracing::error!("{}", message);
```

Comment:
This seems surprising to me. The mechanism is to notify the client of an error, but I don't necessarily want to see errors in my server logs.

Reply:
Maybe debug then. The C++ server does something similar.